cellarium/ml/layers/normadd.py

```diff
@@ -0,0 +1,48 @@
+# Copyright Contributors to the Cellarium project.
+# SPDX-License-Identifier: BSD-3-Clause
+
+from collections.abc import Callable
+
+import torch
+from torch import nn
+
+
+class NormAdd(nn.Module):
+    """
+    Pre-norm layer where the layer normalization is applied before the sublayer.
```

Comment:
This is interesting... because my intuition would be to perform normalization after sublayer _and_ residual, just so that the hidden state remains normalized...

Reply:
I guess one could imagine a scenario that the self-attention layer drives the embeddings to extreme values, and post-sublayer normalization would essentially kill everything else and only keep those extreme values. The pre-norm layer prevents this, by making sure you're adding two normalized values. Perhaps one could even entertain mixing both:
```
norm(z + sublayer(norm(z)))
```

Reply:
Yes, the pre-norm is the standard now.

Yang et al says

> In addition, while our results on width, batch size, sequence length, and training time
still hold for post-layernorm (Fig. 17),11 the transfer across depth only works for pre-layernorm
Transformer

> in fact, post-layernorm Transformers are much more sensitive to HPs than pre-layernorm, so our technique
is more crucial for them, especially for transfer across width. Fig. 1 uses post-layernorm"

Reply:
OLMo uses pre-LN
https://github.com/allenai/OLMo/blob/main/olmo/model.py#L671-L676

HuggingFace GPT2 use pre-LN
https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py#L614-L626

Reply:
Here is an interesting paper about the benefits of the LayerNorm before the attention score computation:

https://arxiv.org/pdf/2305.02582

Reply:
This is the original paper showing that pre-LN is more stable than  post-LN

https://proceedings.mlr.press/v119/xiong20b/xiong20b.pdf